autoscale-groups-nodes-plugin/autoscale-groups-nodes-plugin-core/src/main/java/fish/payara/extensions/autoscale/groups/nodes/NodesScaler.java

```diff
@@ -0,0 +1,405 @@
+                }
+            } else {
+                throw new CommandException(
+                        "Encountered an error scaling up instances. Could not determine instance name");
+            }
+
+            // Adjust the node balance
+            scalingGroupBalance.put(maxNodeEntry.getKey(), maxNodeEntry.getValue() + 1);
+
+            // Increment the counter for our loop
+            instanceCounter++;
+        }
+
+        return instanceNames;
+    }
+
+    /**
+     * Starts the instances in parallel using {@link ScaleCommandHelper}.
+     * @param instanceNames The names of the instances to start
+     * @param actionReport The {@link ActionReport} we want to add out command outputs to
+     */
+    private void startInstances(List<String> instanceNames, ActionReport actionReport) {
+        ScaleCommandHelper scaleCommandHelper = new ScaleCommandHelper(serviceLocator.getService(Domain.class),
+                commandRunner, internalSystemAdministrator.getSubject());
+        scaleCommandHelper.runCommandInParallelAcrossInstances("start-instance", new ParameterMap(),
+                instanceNames, actionReport.addSubActionsReport());
+    }
+
+    @Override
+    public ActionReport scaleDown(int numberOfInstancesToRemove, ScalingGroup scalingGroup) {
+        ActionReport actionReport = commandRunner.getActionReport("plain");
+        try {
+            validate(numberOfInstancesToRemove, scalingGroup);
+        } catch (CommandValidationException commandValidationException) {
+            actionReport.setMessage("Scale down operation cancelled: an error was encountered during validation");
+            actionReport.setFailureCause(commandValidationException);
+            actionReport.setActionExitCode(ActionReport.ExitCode.FAILURE);
+            return actionReport;
+        }
+
+        // Determine which instances to stop, attempting to keep the nodes balanced
+        List<String> instanceNames = determineInstancesToStop(numberOfInstancesToRemove, scalingGroup);
+        // Stop the instances in parallel
+        stopInstances(instanceNames, actionReport.addSubActionsReport());
+        // Delete the instances sequentially
+        deleteInstances(instanceNames, actionReport.addSubActionsReport());
+
+        return actionReport;
+    }
+
+    /**
+     * Determines the names of which instances to stop in the scaling group, attempting to keep the number of instances
+     * on nodes balanced.
+     *
+     * @param numberOfInstancesToRemove The number of instances to remove
+     * @param scalingGroup The scaling group to remove the instances from
+     * @return A list of names of the instances to stop
+     */
+    private List<String> determineInstancesToStop(int numberOfInstancesToRemove, ScalingGroup scalingGroup) {
+        List<String> instanceNames = new ArrayList<>();
+
+        // Quick check: will we just be removing all instances? If so we can skip trying to figure out the balance
```

Comment:
Shouldn't the server prevent all instances to be removed? If we don't worry about this, it might be good to print a warning that at least makes the user aware that the deployment group went empty after the downscale

Reply:
I would say no the server shouldn't prevent you from scaling the deployment group down completely.
Re: warning a user and potentially asking for approval: Enhancement, I'll raise a Jira.